src/tmpfiles/tmpfiles.c

```diff
@@ -1112,6 +1112,8 @@ static int item_do_children(Item *i, const char *path, action_t action) {
                 if (q < 0 && q != -ENOENT && r == 0)
                         r = q;
 
+                dirent_ensure_type(d, de);
+
                 if (IN_SET(de->d_type, DT_UNKNOWN, DT_DIR)) {
```

Comment:
hmm, what precisely is this addition supposed to fix? note that if DT_UNKNOWN is set, we'll actually handle that here, and then will descend into it assuming it is a dir. And if you look above, you'll see we actually handle ENOTDIR on the opendir() just fine. Hence, what is this supposed to fix?

Reply:
I had below entries in my `/etc/tmpfiles.d/logcfg.conf`
```
d   /mnt/storage/logcfg                         2770    logcfg  logaccess -
L   /mnt/storage/logcfg/current                 -       -       -         - /usr/share/defaultlogcfg
Z   /mnt/storage/logcfg                         ~2750   logcfg  logaccess -
```

If entry with unknown type is a symbolic link `/mnt/storage/logcfg/current` indicating to a folder `/usr/share/defaultlogcfg` then errno is `ELOOP` (at least in my case) and below condition is not fulfilled.

```
d = opendir_nomod(path);
    if (!d)
        return IN_SET(errno, ENOENT, ENOTDIR) ? 0 : -errno;
```

Reply:
i see. but i think it would be nicer to then simply also make it handle ELOOP silently, i.e. just add ELOOP to the IN_SET() expression?

Reply:
Handling only ELOOP will help only in such case as mine. 
Ensuring entry type:
- shall cover more cases (if any exist...)
- will be inline with readdir manual:
>       If the file type could not be determined, the value DT_UNKNOWN is returned in d_type.
>
>       Currently, only some filesystems (among them: Btrfs, ext2, ext3, and ext4) have full support for returning the file type in d_type.  All applications must properly handle a return of DT_UNKNOWN.
- is inline with other parts of systemd (see: `shared/install.c`, delta/delta.c)

It is up to you. I can refactor it if you prefer, that's no problem at all.

Reply:
so, the thing is that adding ELOOP has one major benefit: item_do_children() works for the top-level file system object the same way as for any inside, and that's a good thing. With your stat-based solution item_do_children() invoked top-level on a symlink would still fail. And I like generic code... 

Hence, please rework the patch! thanks!

Reply:
that's a good point - a like generic code too.
Ihave reworked it the way you want. However I have a feeling that this does not cover all corner cases... Hope, I'm wrong.